src/ol/geom/flat/textpath.js

```diff
@@ -34,74 +34,114 @@ export function drawTextOnPath(
+  }
+  interpolate = (endM - segmentM) / segmentLength;
+  const endX = lerp(x1, x2, interpolate);
+  const endY = lerp(y1, y2, interpolate);
 
   // Keep text upright
   let reverse;
   if (rotation) {
-    const rotatedCoordinates = rotate(
-      flatCoordinates,
-      offset,
-      end,
-      stride,
-      rotation,
-      [flatCoordinates[offset], flatCoordinates[offset + 1]]
-    );
-    reverse =
-      rotatedCoordinates[0] >
-      rotatedCoordinates[rotatedCoordinates.length - stride];
+    const flat = [beginX, beginY, endX, endY];
+    rotate(flat, 0, 4, 2, rotation, flat, flat);
+    reverse = flat[0] > flat[2];
   } else {
-    reverse = flatCoordinates[offset] > flatCoordinates[end - stride];
+    reverse = beginX > endX;
   }
 
-  const numChars = text.length;
+  const PI = Math.PI;
+  const result = [];
+  const singleSegment = startOffset + stride === offset;
 
-  let x1 = flatCoordinates[offset];
-  let y1 = flatCoordinates[offset + 1];
-  offset += stride;
-  let x2 = flatCoordinates[offset];
-  let y2 = flatCoordinates[offset + 1];
-  let segmentM = 0;
-  let segmentLength = Math.sqrt(Math.pow(x2 - x1, 2) + Math.pow(y2 - y1, 2));
-  let angleChanged = false;
+  offset = startOffset;
+  segmentLength = 0;
+  segmentM = startLength;
+  x2 = flatCoordinates[offset];
+  y2 = flatCoordinates[offset + 1];
+
+  // All on the same segment
+  if (singleSegment) {
+    advance();
 
-  let index, previousAngle;
-  for (let i = 0; i < numChars; ++i) {
-    index = reverse ? numChars - i - 1 : i;
-    const char = text[index];
-    const charLength = scale * measureAndCacheTextWidth(font, char, cache);
-    const charM = startM + charLength / 2;
-    while (offset < end - stride && segmentM + segmentLength < charM) {
-      x1 = x2;
-      y1 = y2;
-      offset += stride;
-      x2 = flatCoordinates[offset];
-      y2 = flatCoordinates[offset + 1];
-      segmentM += segmentLength;
-      segmentLength = Math.sqrt(Math.pow(x2 - x1, 2) + Math.pow(y2 - y1, 2));
+    let previousAngle = Math.atan2(y2 - y1, x2 - x1);
+    if (reverse) {
+      previousAngle += previousAngle > 0 ? -PI : PI;
     }
-    const segmentPos = charM - segmentM;
+    const x = (endX + beginX) / 2;
+    const y = (endY + beginY) / 2;
+    result[0] = [x, y, (endM - startM) / 2, previousAngle, text];
+    return result;
+  }
+
+  let previousAngle;
+  for (let i = 0, ii = text.length; i < ii; ) {
+    advance();
     let angle = Math.atan2(y2 - y1, x2 - x1);
     if (reverse) {
-      angle += angle > 0 ? -Math.PI : Math.PI;
+      angle += angle > 0 ? -PI : PI;
     }
     if (previousAngle !== undefined) {
       let delta = angle - previousAngle;
-      angleChanged = angleChanged || delta !== 0;
-      delta +=
-        delta > Math.PI ? -2 * Math.PI : delta < -Math.PI ? 2 * Math.PI : 0;
+      delta += delta > PI ? -2 * PI : delta < -PI ? 2 * PI : 0;
       if (Math.abs(delta) > maxAngle) {
         return null;
       }
     }
     previousAngle = angle;
```

Comment:
I would have preferred this line to be after the if block with the `continue`  below to measure the angle change between characters instead of the angle change of each line segment. When there are multiple angle changes between two successive characters they can exceed the configured `maxAngle`, but this cannot be taken into account when the line sub segment is determined where the text should be placed here:
https://github.com/openlayers/openlayers/blob/f102e11bbe8304a9c8efa8caa78eb9e630a98daa/src/ol/geom/flat/straightchunk.js#L14